crates/environ/src/func_environ.rs

```diff
@@ -646,27 +687,314 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
+                // * and finally increment the `next` bump finger.
+                builder.switch_to_block(no_gc_block);
+                self.mutate_extenref_ref_count(builder, elem, 1);
+                builder.ins().store(ir::MemFlags::trusted(), elem, next, 0);
+
+                let new_next = builder
+                    .ins()
+                    .iadd_imm(next, i64::from(reference_type.bytes()));
+                builder.ins().store(
+                    ir::MemFlags::trusted(),
+                    new_next,
+                    activations_table,
+                    i32::try_from(self.offsets.vm_extern_ref_activation_table_next()).unwrap(),
+                );
+
+                builder.ins().jump(continue_block, &[]);
+                builder.switch_to_block(continue_block);
+
+                builder.seal_block(non_null_elem_block);
+                builder.seal_block(gc_block);
+                builder.seal_block(no_gc_block);
+                builder.seal_block(continue_block);
+
+                Ok(elem)
+            }
+            ty => Err(WasmError::Unsupported(format!(
+                "unsupported table type for `table.get` instruction: {:?}",
+                ty
+            ))),
+        }
     }
 
     fn translate_table_set(
         &mut self,
-        _: cranelift_codegen::cursor::FuncCursor<'_>,
-        _: TableIndex,
-        _: ir::Table,
-        _: ir::Value,
-        _: ir::Value,
+        builder: &mut FunctionBuilder,
+        table_index: TableIndex,
+        table: ir::Table,
+        value: ir::Value,
+        index: ir::Value,
     ) -> WasmResult<()> {
-        Err(WasmError::Unsupported(
-            "the `table.set` instruction is not supported yet".into(),
-        ))
+        let pointer_type = self.pointer_type();
+
+        let plan = &self.module.table_plans[table_index];
+        match plan.table.wasm_ty {
+            WasmType::FuncRef => match plan.style {
+                TableStyle::CallerChecksSignature => {
+                    let table_entry_addr = builder.ins().table_addr(pointer_type, table, index, 0);
+                    builder
+                        .ins()
+                        .store(ir::MemFlags::trusted(), value, table_entry_addr, 0);
+                    Ok(())
+                }
+            },
+            WasmType::ExternRef => {
+                // Our write barrier for `externref`s being copied out of the
+                // stack and into a table is roughly equivalent to the following
+                // pseudocode:
+                //
+                // ```
+                // if value != null:
+                //     value.ref_count += 1
+                // let current_elem = table[index]
+                // if current_elem != null:
+                //     current_elem.ref_count -= 1
+                //     if current_elem.ref_count == 0:
+                //         call drop_externref_in_place(current_elem)
+                // table[index] = value
```

Comment:
I think technically this is memory unsafe since the destructor can run arbitrary code which can poke at the table. Perhaps the third statement could be `let current_elem = replace(&mut table[index], value)`, and I think that'd fix any issues here?

Reply:
This whole thing is sort of an inline destructor, the `table[index] = ...` bit of pseudocode isn't intended to convey that we are running destructors on the old `table[index]` within this assignment.